mcode/mcode-extraction-framework
# CSV Validation

# Summary

Introduces the concept of validating CSV schemas for the specific extractor (e.g. column names, number of columns, required columns, etc.).

CSV Validation is done at initialization time of the Extractor (before any data is extracted for anything).

**NOTE**: Schemas in this PR only exist for the ICARE data elements per the task definition of done.

## New behavior

If any validation errors occur, they will be logged to the console and the extraction process will stop. Otherwise, logs will indicate that the validation has succeeded, and extraction will continue. Actual extraction behavior is unchanged

## Code changes

* Schemas and utilities for doing validation of specific CSV content based on a provided schema (see `src/helpers/schemas/csv.js`) for schemas. They are simple for now (just column names and optional vs. not). We can add more complex validation in the future but IMO that should be separate tasking.
* Introduce `BaseCSVExtractor` class that all CSV extractors extend from
  * Implement the validate function and also construct the CSV module based on the file path
* Validate all extractors in the `initializeExtractors` in BaseClient
  * Validation is asynchronous, so this is now an async function
* Update all tests with necessary async keywords based on calls to extractor initialization.
* Update the CSV data element spreadsheet to correctly indicate present fields and if they are required based on the behavior in the current version of the MEF.

# Testing guidance

* Running extraction with existing CSV should have unchanged behavior apart from logs
* Change some of the data in any of the ICARE CSVs (Patient, Condition, TreatmentPlanChange, CancerDiseaseStatus, ClinicalTrialInformation)
  * Try omitting required data (e.g. mrn)
  * Try adding erroneous columns
  * Try invalid CSV (e.g. 9 column headers but only 8 column values for a given row)
